src/ssl.ml

```diff
@@ -301,26 +306,137 @@ end
+    let err = get_error socket ret in
+    if err != Error_none then
+      raise (Accept_error err)
+
+  external raw_write : socket -> Bytes.t -> int -> int -> int
+    = "ocaml_ssl_write_blocking" [@@noalloc]
+
+  let write socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.write: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.write: length negative";
+    if start + length > Bytes.length buffer then
+      invalid_arg "Ssl.write: Buffer too short";
+    let ret = raw_write socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Write_error err);
+      end;
+    ret
+
+  external raw_write_substring : socket -> string -> int -> int -> int
+    = "ocaml_ssl_write_blocking" [@@noalloc]
+
+  let write_substring socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.write_substring: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.write_substring: length negative";
+    if start + length > String.length buffer then
+      invalid_arg "Ssl.write_substring: Buffer too short";
+    let ret = raw_write_substring socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Write_error err);
+      end;
+    ret
+
+  external raw_write_bigarray : socket -> bigarray -> int -> int -> int
+    = "ocaml_ssl_write_bigarray_blocking" [@@noalloc]
+
+  let write_bigarray socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.write_bigarray: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.write_bigarray: length negative";
+    if start + length > Bigarray.Array1.dim buffer then
+      invalid_arg "Ssl.write_bigarray: Buffer too short";
+    let ret = raw_write_bigarray socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Write_error err);
+      end;
+    ret
+
+  external raw_read : socket -> Bytes.t -> int -> int -> int
+    = "ocaml_ssl_read_blocking" [@@noalloc]
+
+  let read socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.read: start negative";
+    if length < 0 then
+      invalid_arg "Ssl.read: length negative";
+    if start + length > Bytes.length buffer then
+      invalid_arg "Buffer too short";
+    let ret = raw_read socket buffer start length in
+    if ret <= 0 then
+      begin
+        let err = get_error socket ret in
+        if err != Error_none then
+          raise (Read_error err);
+      end;
+    ret
+
+  external raw_read_into_bigarray :
+    socket -> bigarray -> int -> int -> int =
+    "ocaml_ssl_read_into_bigarray_blocking" [@@noalloc]
+
+  let read_into_bigarray socket buffer start length =
+    if start < 0 then
+      invalid_arg "Ssl.read: start negative";
```

Comment:
```suggestion
      invalid_arg "Ssl.Runtime_lock.read_into_bigarray: start negative";
```

Reply:
This and the later comments should be treated in PR #118, because I had to rebase this PR and #118, and the name are introduced there ?

Reply:
I don't think so, PR #118 doesn't introduce these new functions

Reply:
You are right. I fixed one but did not add the submodule name:
- it is too long
- and I usually do 
``` module Ssl = struct include Ssl include Runtime_lock end ```
- This is even more true if we add the Labels submodule you suggest.

Reply:
In any case, the error message should point to the actual function that is running.

Reply:
Fixed by [d45ce08](https://github.com/savonet/ocaml-ssl/pull/117/commits/d45ce084b4c4a22896155225fb30a153abdf898d)